Feature/syal 1065

ThanhD-AVISPL opened the pull request:
Pull request on SYAL-1065 and SYAL-1066 for Sembient aggregator adapter.

ThanhD-AVISPL committed:
SYAL-1065

Update aggregator and aggregated devices code.

ThanhD-AVISPL committed:
Create SembientAggregatorCommunicator.java

ThanhD-AVISPL committed:
SYAL-1066 Update aggregated device code and UTs

ThanhD-AVISPL committed:
Refactor code

ThanhD-AVISPL committed:
Update controlProperty method

ThanhD-AVISPL committed:
Add UT for polling interval

ThanhD-AVISPL committed:
Refactor code

Fix pmd checks, upgrade api version from 5.6 to 5.7

ThanhD-AVISPL committed:
update more detail Javadoc to DTO classes

ThanhD-AVISPL committed:
Refactor code and uncomment No data UTs

ThanhD-AVISPL committed:
Update SembientAggregatorCommunicator class

ThanhD-AVISPL committed:
Set timeout value to 200 ms

ThanhD-AVISPL committed:
Fix code after internal review

ThanhD-AVISPL committed:
Adjust retry on doDelete and doPut

ThanhD-AVISPL committed:
Remove un-use retry on doPut and doDelete

ThanhD-AVISPL committed:
Update initialize aggregatedDeviceTagMap

ThanhD-AVISPL committed:
Populate tag dropdown after create and delete region tag from cache

ThanhD-AVISPL committed:
Update one more case when creating tag

ThanhD-AVISPL committed:
Update cases when user create all space characters

ThanhD-AVISPL committed:
Rename some properties and fix UT

ThanhD-AVISPL committed:
Update code after latest update on HLD

- Split aggregated device to two types
- Add configurable properties, installationLayoutPollingCycle and pollingCycle

ThanhD-AVISPL committed:
Reformat the average values

ThanhD-AVISPL committed:
Update code to handle case set sensor properties from cached

ThanhD-AVISPL committed:
Fix failed UT, update exception.

ThanhD-AVISPL committed:
Merge pull request #4 from AVISPL/feature/SYAL-1066

Feature/syal 1066

ThanhD-AVISPL committed:
Handle spaces for filtering options and fix UT to match new code.

ThanhD-AVISPL committed:
Fix proper return after populating no data property

ThanhD-AVISPL committed:
Update case where polling cycles are invalid

ThanhD-AVISPL committed:
Update code based on comments

ThanhD-AVISPL committed:
Update no data case

MaksimR-AVISPL commented on pom.xml:
```diff
@@ -0,0 +1,217 @@
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-expression</artifactId>
+            <version>5.2.6.RELEASE</version>
```
The latest version is 5.2.22.RELEASE now, can you please update?

ThanhD-AVISPL replied:
Updated

MaksimR-AVISPL commented on pom.xml:
```diff
@@ -0,0 +1,217 @@
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-web</artifactId>
+            <version>5.2.6.RELEASE</version>
```
The version is also 5.2.22.RELEASE

MaksimR-AVISPL commented on pom.xml:
```diff
@@ -0,0 +1,217 @@
+        <dependency>
+            <groupId>net.java.dev.jna</groupId>
+            <artifactId>jna-platform</artifactId>
+            <version>5.9.0</version>
```
this one is 5.12.0 now

MaksimR-AVISPL commented on pom.xml:
```diff
@@ -0,0 +1,217 @@
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>30.1.1-jre</version>
```
31.1-jre

MaksimR-AVISPL commented on src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java:
```diff
@@ -0,0 +1,1813 @@
+			AggregatedDevice deviceToBeControlled = aggregatedDevices.get(deviceId);
+			String[] deviceDetails = deviceId.split(SembientAggregatorConstant.DASH);
+			int lastIndex = deviceDetails.length - 1;
+			String deviceName = deviceDetails[lastIndex];
```
do you think deviceDetails format should be validated here? Just in case something goes wrong and wrong Id is populated, since the format is a bit complex there

ThanhD-AVISPL replied:
Sure, let me do some validation here.

MaksimR-AVISPL commented on src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java:
```diff
@@ -0,0 +1,1813 @@
+				aggregatedDeviceHourMap.put(deviceId, (String) controllableProperty.getValue());
+				OccupancyData[] occupancyData = aggregatedDeviceOccupancyMap.get(deviceId);
+				if (occupancyData == null) {
+					throw new ResourceNotReachableException("Fail to control OccupancyList, Hour dropdown.");
```
The exception type doesn't seem to fit the best this use case. the reason is that there's no data in aggregatedDeviceOccupancyMap, so maybe this is some kind of "IllegalArgument" or maybe just a generic RuntimeException. Depends on the reason occupancyData is not here - if this is the deviceId issue, or that something is off with the data integrity in the map

MaksimR-AVISPL commented on src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java:
```diff
@@ -0,0 +1,1813 @@
+						} catch (CommandFailureException e) {
+							logger.error("Fail to create with status code: " + e.getStatusCode() + ", value: " + newTag, e);
+							if (e.getStatusCode() == 429) {
+								throw new IllegalStateException("Too many request, please try to create region tag with value: " + newTag + " later.");
```
I think it'd sound a bit more smooth with "Too many request**s**"

MaksimR-AVISPL commented on src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java:
```diff
@@ -0,0 +1,1813 @@
+							if (e.getStatusCode() == 429) {
+								throw new IllegalStateException("Too many request, please try to create region tag with value: " + newTag + " later.");
+							} else {
+								throw new IllegalStateException("Fail to create region tag with value: " + newTag);
```
and here - "Fail**ed** to create ..."
I just think it'd sound a bit nicer :)

MaksimR-AVISPL commented on src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java:
```diff
@@ -0,0 +1,1813 @@
+								throw new IllegalStateException("Fail to create region tag with value: " + newTag);
+							}
+						} catch (Exception e) {
+							logger.error("Exception occur when creating region tag with value: " + newTag, e);
```
"occur**red**", since the exception has already happened :)

MaksimR-AVISPL commented on src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java:
```diff
@@ -0,0 +1,1813 @@
+						}
+						if (createRegionTagWrapperControl != null) {
+							if (!SembientAggregatorConstant.STATUS_CODE_200.equals(createRegionTagWrapperControl.getStatusCode())) {
+								throw new IllegalStateException("Fail to create region with value is: " + controllableProperty.getValue());
```
also "Fail**ed**" here. Could you please the rest of the the error messages?

MaksimR-AVISPL commented on src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java:
```diff
@@ -0,0 +1,1813 @@
+							}
+						} catch (Exception e) {
+							logger.error("Exception occur when deleting region tag with value: " + valueToBeDelete, e);
+							throw new IllegalStateException("Fail to delete region tag with value: " + valueToBeDelete);
```
instead of throwing an illegal exception here - could you just throw the original one here? Because in this block all that happens is only an exception rethrow

ThanhD-AVISPL replied:
I rethrown it as IllegalStateException here is because I want the error message much clearer to the user, do you think it makes sense? Because if we throw original one here, it might hard to read.

MaksimR-AVISPL commented on src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java:
```diff
@@ -0,0 +1,1813 @@
+	private void populateRegionTag(Map<String, String> properties, List<AdvancedControllableProperty> controls, String deviceId) throws Exception {
+		controls.add(createText(properties, SembientAggregatorConstant.REGION_TAG_NEW_TAG, lastNewTag.get(deviceId)));
+		controls.add(createButton(properties, SembientAggregatorConstant.REGION_TAG_CREATE, SembientAggregatorConstant.LABEL_CREATE, SembientAggregatorConstant.LABEL_PRESSED_CREATING));
+		String[] regionDetails = deviceId.split(SembientAggregatorConstant.DASH);
```
similarly to controllable properties operation - do you think the deviceId should be validated at this point, since there's a specific pattern that used for the deviceId creation

ThanhD-AVISPL replied:
I don't think we should validate it here. Because if we reach this method, it's the latest populated device ID.

ThanhD-AVISPL committed:
Fix code after code review.

1. Update dependencies version.
2. Update error logs, exception types and validation for deviceId when control.
3. Update failed UTs

ThanhD-AVISPL committed:
Update deprecated github action

ThanhD-AVISPL committed:
Change pollingCycle default value to and refactor code
